src/app_service/service/provider/service.nim

```diff
@@ -0,0 +1,324 @@
+
+logScope:
+  topics = "provider-service"
+
+const HTTPS_SCHEME* = "https"
+const IPFS_GATEWAY* =  ".infura.status.im"
+const SWARM_GATEWAY* = "swarm-gateways.net"
+
+type
+  RequestTypes {.pure.} = enum
+    Web3SendAsyncReadOnly = "web3-send-async-read-only",
+    HistoryStateChanged = "history-state-changed",
+    APIRequest = "api-request"
+    Unknown = "unknown"
+
+  ResponseTypes {.pure.} = enum
+    Web3SendAsyncCallback = "web3-send-async-callback",
+    APIResponse = "api-response",
+    Web3ResponseError = "web3-response-error"
+
+type
+  Payload = ref object
+    id: JsonNode
+    rpcMethod: string
+
+  Web3SendAsyncReadOnly = ref object
+    messageId: JsonNode
+    payload: Payload
+    request: string
+    hostname: string
+
+  APIRequest = ref object
+    isAllowed: bool
+    messageId: JsonNode
+    permission: Permission
+    hostname: string
+
+const AUTH_METHODS = toHashSet(["eth_accounts", "eth_coinbase", "eth_sendTransaction", "eth_sign", "keycard_signTypedData", "eth_signTypedData", "eth_signTypedData_v3", "personal_sign", "personal_ecRecover"])
+const SIGN_METHODS = toHashSet(["eth_sign", "personal_sign", "eth_signTypedData", "eth_signTypedData_v3"])
+const ACC_METHODS = toHashSet(["eth_accounts", "eth_coinbase"])
+
+proc requestType(message: string): RequestTypes =
+  let data = message.parseJson
+  result = RequestTypes.Unknown
+  try:
+    result = parseEnum[RequestTypes](data["type"].getStr())
+  except:
+    warn "Unknown request type received", value=data["permission"].getStr()
+
+
+proc toWeb3SendAsyncReadOnly(message: string): Web3SendAsyncReadOnly =
+  let data = message.parseJson
+  result = Web3SendAsyncReadOnly(
+    messageId: data["messageId"],
+    request: $data["payload"],
+    hostname: data{"hostname"}.getStr(),
+    payload: Payload(
+      id: data["payload"]{"id"},
+      rpcMethod: data["payload"]["method"].getStr()
+    )
+  )
+
+proc toAPIRequest(message: string): APIRequest =
+  let data = message.parseJson
+
+  result = APIRequest(
+    messageId: data["messageId"],
+    isAllowed: data{"isAllowed"}.getBool(),
+    permission: data["permission"].getStr().toPermission(),
+    hostname: data{"hostname"}.getStr()
+  )
+
+
+
+type 
+  Service* = ref object of service_interface.ServiceInterface
+    dappPermissionsService: dapp_permissions_service.ServiceInterface
+    settingsService: settings_service.ServiceInterface
+    ensService: ens_service.ServiceInterface
+
+method delete*(self: Service) =
+  discard
+
+proc newService*(dappPermissionsService: dapp_permissions_service.ServiceInterface, 
+    settingsService: settings_service.ServiceInterface,
+    ensService: ens_service.ServiceInterface): Service =
+  result = Service()
+  result.dappPermissionsService = dappPermissionsService
+  result.settingsService = settingsService
+  result.ensService = ensService
+
+method init*(self: Service) =
+  discard
+
+proc process(self: Service, data: Web3SendAsyncReadOnly): string =
```

Comment:
We have any chance that `data` can be a null?

Reply:
Unrelated, but this function is pretty big. Maybe we could split it in sub-functions for each rpcMethod.